contracts/IdentityV1.sol

```diff
@@ -0,0 +1,277 @@
+            executions[_id].approved = true;
+            success = executions[_id].to.call(executions[_id].data, 0);
+            if (success) {
+                executions[_id].executed = true;
+                emit Executed(
+                    _id,
+                    executions[_id].to,
+                    executions[_id].value,
+                    executions[_id].data
+                );
+                return;
+            } else {
+                emit ExecutionFailed(
+                    _id,
+                    executions[_id].to,
+                    executions[_id].value,
+                    executions[_id].data
+                );
+                return;
+            }
+        } else {
+            executions[_id].approved = false;
+        }
+        return true;
+    }
+
+    function execute(address _to, uint256 _value, bytes _data)
+        public
+        returns (uint256 executionId)
+    {
+        require(!executions[executionNonce].executed, "Already executed");
+        executions[executionNonce].to = _to;
+        executions[executionNonce].value = _value;
+        executions[executionNonce].data = _data;
+
+        emit ExecutionRequested(executionNonce, _to, _value, _data);
+
+        if (keyHasPurpose(bytes32(msg.sender), ACTION_KEY)) {
+            approve(executionNonce, true);
+        }
+
+        executionNonce++;
+        return executionNonce-1;
+    }
+
+    function removeKey(bytes32 _key, uint256 _purpose)
+        public
+        onlyManagement
+        returns (bool success)
+    {
+        require(keys[_key].key == _key, "No such key");
+
+        if (!keyHasPurpose(_key, _purpose)) {
+            return false;
+        }
+
+        uint256 arrayLength = keys[_key].purpose.length;
+        int index = -1;
+        for (uint i = 0; i < arrayLength; i++) {
+            if (keys[_key].purpose[i] == _purpose) {
+                index = int(i);
+                break;
+            }
+        }
+
+        if (index != -1) {
+            keys[_key].purpose[uint(index)] = keys[_key].purpose[arrayLength - 1];
+            delete keys[_key].purpose[arrayLength - 1];
+            keys[_key].purpose.length--;
+        }
+
+        uint256 purposesLen = keysByPurpose[_purpose].length;
+        for (uint j = 0; j < purposesLen; j++) {
+            if (keysByPurpose[_purpose][j] == _key) {
+                keysByPurpose[_purpose][j] = keysByPurpose[_purpose][purposesLen - 1];
+                delete keysByPurpose[purposesLen - 1];
+                keysByPurpose[_purpose].length--;
+                break;
+            }
+        }
+
+        emit KeyRemoved(_key, _purpose, keys[_key].keyType);
+
+        return true;
+    }
+
+    function keyHasPurpose(bytes32 _key, uint256 _purpose)
+        public
+        view
+        returns(bool result)
+    {
+        if (keys[_key].key == 0) return false;
+        uint256 arrayLength = keys[_key].purpose.length;
+        for (uint i = 0; i < arrayLength; i++) {
```

Comment:
Probably... let me see

Reply:
That could work, however there will still be a for loop - just iterating over a different array. It is unknown which one will be longer - it depends on how the owner of the identity uses it.

Reply:
I was thinking of optimisations, (NOT FOR NOW but heres the idea, you create another mapping which takes a key and returns the index in the array, this will increase storage but runtime execeution will be fast

Reply:
In my experiments iterations over an array were coming out a lot cheaper than storage. In reality this array will only have a handful of elements so it is already quite cheap to use.